cmd/expired-authz-purger/main.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/cactus/go-statsd-client/statsd"
+	"github.com/jmhodges/clock"
+	"gopkg.in/gorp.v1"
+
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+type expiredAuthzPurger struct {
+	stats statsd.Statter
+	log   blog.Logger
+	clk   clock.Clock
+	db    *gorp.DbMap
+
+	batchSize int64
+}
+
+func (p *expiredAuthzPurger) purgeAuthzs(purgeBefore time.Time, yes bool) (int64, error) {
+	if !yes {
+		var count int
+		err := p.db.SelectOne(&count, `SELECT COUNT(1) FROM pendingAuthorizations AS pa WHERE expires <= ?`, purgeBefore)
+		if err != nil {
+			return 0, err
+		}
+		reader := bufio.NewReader(os.Stdin)
+		for {
+			fmt.Fprintf(os.Stdout, "\nAbout to purge %d pending authorizations, proceed? [y/N]: ", count)
+			text, err := reader.ReadString('\n')
+			if err != nil {
+				return 0, err
+			}
+			text = strings.ToLower(text)
+			if text != "y\n" && text != "n\n" && text != "\n" {
+				continue
+			}
+			if text == "n\n" || text == "\n" {
+				os.Exit(0)
+			} else {
+				break
+			}
+		}
+	}
+
+	rowsAffected := int64(0)
+	for {
+		result, err := p.db.Exec(`
+			DELETE FROM pendingAuthorizations
+			WHERE expires <= ?
+			LIMIT ?
+			`,
+			purgeBefore,
+			p.batchSize,
+		)
+		if err != nil {
+			return rowsAffected, err
+		}
+		rows, err := result.RowsAffected()
+		if err != nil {
+			return rowsAffected, err
+		}
+
+		p.stats.Inc("PendingAuthzDeleted", rows, 1.0)
+		rowsAffected += rows
+		p.log.Info(fmt.Sprintf("Progress: Deleted %d (%d total) expired pending authorizations", rows, rowsAffected))
+
+		if rows < p.batchSize {
+			p.log.Info(fmt.Sprintf("Deleted a total of %d expired pending authorizations", rowsAffected))
+			return rowsAffected, nil
+		}
+	}
+}
+
+func main() {
+	batchSize := flag.Int("batch-size", 1000, "Size of batches to do SELECT queries in")
+	force := flag.Bool("force", false, "Allows purge of all expired pending authorizations (dangerous)")
```

Comment:
This flag seems unnecessary. If someone wants to purge "all" expired authz they can just pass a trivial grace period, like 1 microsecond.